src/main/scala/org/wikimedia/analytics/hdfstools/HdfsRsyncExec.scala

```diff
@@ -378,49 +350,141 @@ class HdfsRsyncExec(config: HdfsRsyncConfig) {
+                Try(config.srcFs.globStatus(src).toSeq).getOrElse(Seq.empty)
+                    .map(s => (s, Some(s.getPath.getParent)))
+            } else {
+                // In a recursion folder, use listStatus and keep basePath unchanged
+                config.srcFs.listStatus(src).map(s => (s, basePath))
+            }
+        }).groupBy { case (fileStatus, basePath) => fileStatus.getPath.getName }
+    }
+
+    /**
+     * Function listing the files and folders from a dstPath. Result is organised in a
+     * map of (filename, filesStatus)
+     * @param dstPath the dst path to scan
+     * @return the map of fileStatuses keyed by filenames.
+     */
+    private def getDstFilesAndFolders(dstPath: Option[Path]): Map[String, FileStatus] = {
+        dstPath.toSeq.flatMap(path => {
+            // We need this check that dst exists for dryrun mode
+            // as dst dirs are not created as we recurse into them
+            if (config.dstFs.exists(path)) {
+                config.dstFs.listStatus(path).map(f => (f.getPath.getName, f))
+            } else {
+                Seq.empty
+            }
+        }).toMap
+    }
+
+    /**
+     * Function filtering a list of source and basePath by applying filter-rules defined in configuration.
+     *
+     * @param srcList the list of (FileStatus, basePath) to filer
+     * @return the filtered list
+     */
+    private def applyFilterRules(srcList: Seq[(FileStatus, Option[Path])]): Seq[(FileStatus, Option[Path])] = {
+        srcList.filter { case (src, basePath) =>
+            val matchingRule = config.parsedFilterRules.find(rule => rule.matches(src, basePath.get))
+            if (matchingRule.isDefined && matchingRule.get.ruleType == Exclude()) {
+                log.debug(s"EXCLUDE_SRC - ${src.getPath}")
+                false
+            } else {
+                true
+            }
+        }
+    }
+
+    /**
+     * This function either processes a folder into target and recurse if all sources are
+     * directories and the recursion flag is on (directories are skipped if flag is off),
+     * or processes a file if it is alone in its list.
+     * If the list contains incoherent objects (both files and folders, or multiple files),
+     * an [[IllegalStateException]] is thrown.
+     *
+     * @param srcList the srcList of files sharing the same filename and target.
+     * @param target the target were to merge folders or copy/update files (if any).
+     * @param foundTarget the found target in dst (if any)
+     */
+    def mergeOrProcessCoherentSrcList(
+        srcList: Seq[(FileStatus, Option[Path])],
+        target: Option[Path],
+        foundTarget: Option[FileStatus]
+    ): Unit = {
+        // If srcList contains only directories
+        if (srcList.forall { case (s, _) => s.isDirectory }) {
+            if (config.recurse) {
+                // Create directory if it doesn't exist and recurse
+                processDir(target, foundTarget)
```

Comment:
You don't use the return value from processDir.  Is processDir only responsible for creating any missing src dirs that are not in dst?  If so, should we rename the processDir function accordingly?

Reply:
Ah, I see, processDir creates dirs the same as in src, but processFile actually does copying.

Reply:
Hm, did you remove the ability for directory copy without recursion? (ok with me if you did)